### PR TITLE
fp16 gradient clip fix

### LINKED ISSUE #184: Batch norm is not compatible with gradient norm clipping in float16 mode

### PR DESCRIPTION
**Pull request by blisc**
Fixes #184. Gradients are converted to float32 before clipping, and then converted back to the correct dtype after.

### CODE REVIEW

File: open_seq2seq/optimizers/optimizers.py [hunk 1]
```diff
@@ -24,9 +24,14 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import collections
 import six
 import tensorflow as tf
+from tensorflow.python.framework import ops
```
Comment (Kipok) [inline open_seq2seq/optimizers/optimizers.py]: I think it is a good practice to rewrite these imports in a usual TF way. For example, "ops.name_scope" should be just "tf.name_scope" and such. They are used in this way because inside the main TF code base that is the only way to do it, but later they add a bunch of convenient import shortcuts that we should use so that we (and everyone else) are not confused by these "unusual" imports.

File: open_seq2seq/optimizers/optimizers.py [hunk 2]
```diff
@@ -361,5 +366,93 @@ def _global_norm_with_cast(grads_and_vars):
...
+      global_norm = sqrt(sum([l2norm(t)**2 for t in t_list]))
+  If `clip_norm > global_norm` then the entries in `t_list` remain as they are,
+  otherwise they're all shrunk by the global ratio.
+  Any of the entries of `t_list` that are of type `None` are ignored.
+  This is the correct way to perform gradient clipping (for example, see
+  [Pascanu et al., 2012](http://arxiv.org/abs/1211.5063)
+  ([pdf](http://arxiv.org/pdf/1211.5063.pdf))).
+  However, it is slower than `clip_by_norm()` because all the parameters must be
+  ready before the clipping operation can be performed.
+  Args:
```
Comment (Kipok) [inline open_seq2seq/optimizers/optimizers.py]: Please, add new lines so that this doc is rendered correctly by Sphinx.
